### PR TITLE
Add WFG benchmark test

### PR DESCRIPTION
**Pull request by kei-mo**
## Motivation
Based on private discussion with @HideakiImamura and @hvy , we're going to add WFG multi-objective optimization benchmark test to `benchmarks/run_mo_kurobako.py`

## Description of the changes

- [x] Confirm how to call a custom problem from `benchmarks/run_mo_kurobako.py`.
- [x]  Create a kurobako WFG benchmark test by using @HideakiImamura 's [code](https://github.com/HideakiImamura/WFGtestSuite) as a reference.
- [x] Test new benchmark with random sampler and/or TPE sampler.

### CODE REVIEW

**Comment by hvy** [comment]
@kei-mo, thanks for the last discussion and swift PR. 🙇
Could you possibly add to the PR description what exact tasks remain, given its still work in progress? 

@sile, not sure how busy you are but would you be able to review this change later on, as the author of Kurobako? 
As for the background, it's documented in https://github.com/optuna/optuna/issues/3245.
In short, the usage of Kurobako has been somewhat streamlined recently https://github.com/optuna/optuna/blob/master/benchmarks/README.md, and we're now integrating multi-objective problems (this PR) and solvers (in addition to single-objective).

**Comment by sile** [comment]
I'm willing to review this PR. However, it would take some time because I'm a bit busy for a while. If you want to merge this PR within a few days or weeks, please re-assign to another person instead of me 🙏

**Comment by hvy** [comment]
I see, thanks for the quick response. @nzw0301 do you happen to be available?

**Comment by kei-mo** [comment]
> @kei-mo, thanks for the last discussion and swift PR. bow Could you possibly add to the PR description what exact tasks remain, given its still work in progress?

@hvy , I've updated the PR description. I'm currently working on transferring @HideakiImamura san's WFG code to kurobako problem. I think this would take some time (maybe a week).

**Comment by kei-mo** [comment]
@HideakiImamura , if you have the script for running the WFG code, it would be very helpful. Could you share if you don't mind?

**Comment by HideakiImamura** [comment]
@kei-mo Thanks for the update! This is the usage of the WFG. Please take a look.
```python
from WFGtestSuite import wfg

import numpy as np
import math


class WFG:
    def __init__(self, problem_name, n_var, n_obj=2, k=2):
        if problem_name == "wfg1":
            self.wfg = wfg.WFG1(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg2":
            self.wfg = wfg.WFG2(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg3":
            self.wfg = wfg.WFG3(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg4":
            self.wfg = wfg.WFG4(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg5":
            self.wfg = wfg.WFG5(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg6":
            self.wfg = wfg.WFG6(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg7":
            self.wfg = wfg.WFG7(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg8":
            self.wfg = wfg.WFG8(n_arguments=n_var, n_objectives=n_obj, k=k)
        elif problem_name == "wfg9":
            self.wfg = wfg.WFG9(n_arguments=n_var, n_objectives=n_obj, k=k)
        else:
            raise ValueError("not found", problem_name)
        self._low = 0
        self._high = 2
        self.n_var = n_var

    def objective(self, trial):
        x = np.array(
            [
                trial.suggest_float(f"x{i}", self._low, self._high * i)
                for i in range(1, self.n_var + 1)
            ]
        )
        v = self.wfg(x)
        if math.isnan(v[0]) or math.isinf(v[0]):
            raise ValueError
        if math.isnan(v[1]) or math.isinf(v[1]):
            raise ValueError
        return v[0], v[1]
```

**Comment by HideakiImamura** [comment]
By the way, you can pass the CI job `Checks` by running `formats.sh`.

File: benchmarks/problems/WFGtestSuite/wfg.py [hunk 1]
```diff
@@ -0,0 +1,706 @@
+from typing import List
+
+import numpy as np
+
+from . import shape_functions
+from . import transformation_functions
```
Comment (HideakiImamura) [inline benchmarks/problems/WFGtestSuite/wfg.py:6]: Please use the absolute import.
```suggestion
from benchmarks.problems.wfg import shape_functions
from benchmarks.problems.wfg import transformation_functions
```

File: benchmarks/problems/binh_and_korn_problem.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from kurobako import problem
```
Comment (HideakiImamura) [inline benchmarks/problems/binh_and_korn_problem.py:1]: I think we can remove this file from this PR.

File: benchmarks/problems/wfg_problem.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import math
```
Comment (HideakiImamura) [inline benchmarks/problems/wfg_problem.py:1]: Cannot we directly use `wfg.py` or fix it to use with kurobako? I think the codes of this file are redundant.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR! I have several early comments. Please take a look.

At first, how about renaming the `WFGtestSuite` directory to `wfg`? It it simple and easy to understand.

**Comment by kei-mo** [comment]
@HideakiImamura 
I'm working on this task as a part of Optuna Sprint 6

**Comment by kei-mo** [comment]
Here's what I got by running TPE on the WFG benchmarks.
@HideakiImamura -san could you take a look at these results?
Please let me know if you need any additional information.

```
pip list | grep kurobako
```
-> 0.2.1


```bash
python benchmarks/run_mo_kurobako.py \
          --path-to-kurobako <path-to-kurobako> \
          --name "performance-benchmarks" \
          --n-runs 1 \
          --n-jobs 4 \
          --sampler-list "RandomSampler TPESampler " \
          --sampler-kwargs-list "{} {}" \
          --seed 0 \
          --data-dir "." \
          --out-dir "out"
```

<details><summary>Optimization results in WFG benchmark tests</summary>
<p>  

WFG1
wfg1-performance-benchmarks-tpesampler-84645d9a0fdbcf6cff7bfd9f753086931eab1917e807169b73fb83d391eaa7e9


WFG2
wfg2-performance-benchmarks-tpesampler-4c87d66fb8befaeb65724cfff5b2ebd716d89db68d1789aa6a0e805072e94b37


WFG3
wfg3-performance-benchmarks-tpesampler-1c15c96150a573bf04f022589ddbd59e6cea49a252ad07aebc266f5c0fa834ad


WFG4
wfg4-performance-benchmarks-tpesampler-953facdda1af6dd206ce3a45d8714f24ad588b7caf46b98ffcace351160266f5


WFG5
wfg5-performance-benchmarks-tpesampler-9a7eddcd2b60a7de41aade998b90a5dc9f155da8c379762fe67490c10ec21629

WFG6
wfg6-performance-benchmarks-tpesampler-bf249f43ae50ae142375417d1cb0db7574434d3896a2714a48afcbe1d70342e2

WFG7
wfg7-performance-benchmarks-tpesampler-c6a6cbac7ba0f41075190b81713a6f142945b29112509f04a6d4c88f3d774485


WFG8
wfg8-performance-benchmarks-tpesampler-4b29fc819549293ae1af39767dd76f171a3263e77f75f7a9e79abe62d07af881


WFG9
wfg9-performance-benchmarks-tpesampler-41e17c29f942fd00c8c753f93dfb2d250538070673b774a6862a4ae9fa143145

</p>
</details>

**Comment by HideakiImamura** [comment]
The benchmark result basically looks good to me. It might be good to increate the number of trials like 300.

**Comment by kei-mo** [comment]
> @kei-mo Since this PR has a very large number of lines and is expected to take some time to review, why not hold a mob review once to get an overview in real time?

@HideakiImamura Sounds good to me!

**Comment by HideakiImamura** [comment]
@kei-mo Thanks! Please check discord to to a mob-review.

**Review by HideakiImamura — APPROVED** [review]
Thanks, LGTM. I have two memos.

**Comment by HideakiImamura** [comment]
Follow-up: refactor `benchmarks/kurobako/problems/wfg/`

**Review by sile — APPROVED** [review]
I couldn't review the detail of the implementations of WFG functions. But the overall code looks good to me 👍

**Pull request merged**